arrow-fx-coroutines/src/main/kotlin/arrow/fx/coroutines/stream/TerminalOps.kt

```diff
@@ -84,14 +97,29 @@ suspend fun <O> Stream<O>.lastOrNull(): O? =
   foldChunks<O, O?>(null) { acc, c -> c.lastOrNull() ?: acc }
 
 /**
- * Runs all the effects of this [Stream], raising a [NoSuchElementException] if the stream emitted no values,
+ * Runs all the effects of this [Stream], raising a [NoSuchElementException], by default, if the stream emitted no values,
  * and returning the last value emitted otherwise.
  *
  * This a terminal operator, meaning this functions `suspend`s until the [Stream] finishes.
  * If any errors are raised while streaming, it's thrown from this `suspend` scope.
+ *
+ * It's possible to include an alternative [FailStrategy] which describes how to fail in case of not finding the desired element.
+ * //sampleStart
+ * suspend fun main(): Unit =
+ *   Stream(1, 2, 3)
+ *     .drop(3)
+ *     .lastOrError { error("Oops!") } // equivalent to .lastOrError { throw IllegalStateException("Oops!") }
+ *     .let(::println) // not reached, IllegalStateException("Oops!") is thrown
+ * //sampleEnd
+ * ```
  */
-suspend fun <O> Stream<O>.lastOrError(): O =
-  lastOrNull() ?: throw NoSuchElementException()
+suspend fun <O> Stream<O>.lastOrError(failStrategy: FailStrategy = ::noSuchElement): O =
+  lastOrNull() ?: failStrategy()
+
+/**
+ * This typealias defines a function that can only compile if we throw an exception, or call a function that returns [Nothing].
+ */
+typealias FailStrategy = () -> Nothing
```

Comment:
Nit: Seems a bit excessive to create a public `typealias` for this, since we'll expose that `typealias` to all user codebases.

Reply:
Also I think this may be better modeled as `() -> O` and treated as a general fallback rather than an error. This allows a bit more use for those that would directly go to a fallback or catching anyway. `throw` would still work just as well because `Nothing` can be used in place of any type.

Reply:
I was thinking about that while doing this... We could do something like:
```
fun Stream<O>.firstOrElse(block: suspend () -> O)
```
And I would go even further that if we rename `firstOrNull()` to `first()` and we create the nullable operators that match that of Option then we can do:
```kotlin
stream.first().orElse { error("Boom!") }
```

I've been debating with myself whether we want to encourage composition of operations (as good functional programmers) or provide shortcuts like these that do more than one thing.  I guess we could have both and let users decide what to use :) I'm leaning more on the more atomic operations :)